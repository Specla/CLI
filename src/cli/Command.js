import fs from 'fs'
import { resolve } from 'path'
import readline from 'readline'

export default class Command {
  options = {}

  /**
   * Return a template file
   * @param  {String} path
   * @return {String}
   */
  _template (path) {
    return fs.readFileSync(resolve(__dirname, '../../templates', path))
  }

  /**
   * Create file from template
   * @param  {String} file
   * @return {Boolean}
   */
  _createFile (file) {
    return fs.writeFileSync(
      resolve(this.projectPath ||Â process.cwd(), file),
      this._template(file)
    )
  }

  /**
   * Listen for data on stdin
   * @return {Promise}
   */
  stdin () {
    return Promise(resolve => {
      process.stdin.on('readable', () => {
        const data = process.stdin.read()

        if (data !== null) {
          resolve(data)
        }
      })
    })
  }

  /**
   * Write message to stdout
   * @param {String} message
   */
  stdout (message) {
    process.stdout.write(message)
  }

  /**
   * Ask the user a question
   * @param  {String} question
   * @return {Promise}
   */
  prompt (question) {
    return new Promise(resolve => {
      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
      })

      rl.question(question + ' ', result => {
        resolve(result)
        rl.close()
      })
    })
  }

  /**
   * Exit process with an exit code
   * @param  {Number} [code=1]
   */
  exit (code) {
    process.exit(code)
  }
}
