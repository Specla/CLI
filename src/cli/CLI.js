import minimist from 'minimist'
import Specla from '../'
import speclaCommands from './commands'
import Command from './Command'

export default class CLI extends Specla {
  /**
   * CLI arguments
   * @type {Array}
   */
  _args = []

  /**
   * CLI options
   * @type {Object}
   */
  _options = {}

  /**
   * Given command
   * @type {String}
   */
  _command = null

  /**
   * Available commands
   * @type {Object}
   */
  _commands = {}

  /**
   * Create a new Specla CLI instance
   * @return {CLI}
   */
  constructor () {
    super({ 'specla.runtime': 'cli' })
    this._parseMinimistArgs()
    this._registerSpeclaCommands()

    Command.args = this._args
    Command.options = this._options
    Command.availableCommands = this._commands
  }

  /**
   * Parse the minimist arguments to a command, arguments and options
   */
  _parseMinimistArgs () {
    const args = minimist(process.argv.slice(2))
    this._command = args._[0]
    args._.shift()
    this._args = args._
    delete args._
    this._options = args
  }

  /**
   * Register the specla commands
   */
  _registerSpeclaCommands () {
    for (const command of speclaCommands) {
      if (typeof command.signature === 'string') {
        this._commands[command.signature] = command
      }

      if (Array.isArray(command.signature)) {
        for (const signature of command.signature) {
          this._commands[signature] = command
        }
      }
    }
  }

  /**
   * Run CLI command
   * @return {Mixed}
   */
  run () {
    for (const command in this._commands) {
      if (command === this._command) {
        return new this._commands[command](...this._args)
      }

      if (this._command) {
        continue;
      }

      for (const option in this._options) {
        if (
          command === '-' + option ||
          command === '--' + option
        ) {
          return new this._commands[command]()
        }
      }
    }

    return new this._commands['help']() // eslint-disable-line
  }
}
