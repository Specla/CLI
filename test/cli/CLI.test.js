/* eslint-env jest */
import CLI from '../../src/cli/CLI'
import Command from '../../src/cli/Command'

afterEach(() => CLI.reset())

test('Should set runtime to cli', () => {
  const cli = new CLI()
  expect(cli.config.get('specla.runtime')).toBe('cli')
})

test('Should load the specla commands', () => {
  const cli = new CLI()
  expect(cli._commands).toMatchSnapshot()
})

test('Should parse arguments and options', () => {
  const backup = process.argv
  process.argv = ['node', 'specla', 'test:command', 'argument', '-th', '--name=name', '--lastname', 'lastname']
  const cli = new CLI()
  expect(cli._command).toBe('test:command')
  expect(cli._args).toMatchSnapshot()
  expect(cli._options).toMatchSnapshot()
  process.argv = backup
})

test('Should parse argument and options to the Command class', () => {
  const cli = new CLI()
  expect(cli._options).toBe(Command.options)
  expect(cli._args).toBe(Command.args)
})

test('Should trigger run the command', () => {
  const cli = new CLI()
  expect(cli.run()).toMatchSnapshot()
})
