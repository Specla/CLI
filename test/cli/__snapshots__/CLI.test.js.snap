// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should load the specla commands 1`] = `
Object {
  "--help": [Function],
  "--version": [Function],
  "-h": [Function],
  "-v": [Function],
  "build": [Function],
  "create": [Function],
  "create:command": [Function],
  "help": [Function],
  "version": [Function],
}
`;

exports[`Should parse arguments and options 1`] = `
Array [
  "argument",
]
`;

exports[`Should parse arguments and options 2`] = `
Object {
  "h": true,
  "lastname": "lastname",
  "name": "name",
  "t": true,
}
`;

exports[`Should trigger run the command 1`] = `
Help {
  "_commands": Object {
    "--help": [Function],
    "--version": [Function],
    "-h": [Function],
    "-v": [Function],
    "build": [Function],
    "create": [Function],
    "create:command": [Function],
    "help": [Function],
    "version": [Function],
  },
}
`;
